File: backend/app/schemas/sale_entry.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class SaleEntryCreate(BaseModel):
    company_id: UUID | None = None
    invoice_no: str | None = None
    sale_date: date
    buyer_id: UUID
    product_id: UUID
    quantity: Decimal
    rate: Decimal
    gross_amount: Decimal
    transport_cost: Decimal = Decimal("0")
    lr_no: str | None = None
    driver_name: str | None = None
    vehicle_no: str | None = None
    owner_name: str | None = None
    hsn_code: str | None = None
    tcs_amount: Decimal = Decimal("0")
    add_topay: Decimal = Decimal("0")
    less_topay: Decimal = Decimal("0")
    net_amount: Decimal
    po_no: str | None = None
    branch: str | None = None


class SaleEntryUpdate(BaseModel):
    company_id: UUID | None = None
    invoice_no: str | None = None
    sale_date: date | None = None
    buyer_id: UUID | None = None
    product_id: UUID | None = None
    quantity: Decimal | None = None
    rate: Decimal | None = None
    gross_amount: Decimal | None = None
    transport_cost: Decimal | None = None
    lr_no: str | None = None
    driver_name: str | None = None
    vehicle_no: str | None = None
    owner_name: str | None = None
    hsn_code: str | None = None
    tcs_amount: Decimal | None = None
    add_topay: Decimal | None = None
    less_topay: Decimal | None = None
    net_amount: Decimal | None = None
    po_no: str | None = None
    branch: str | None = None


class SaleEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    company_id: UUID | None
    invoice_no: str | None
    sale_date: date
    buyer_id: UUID
    product_id: UUID
    quantity: Decimal
    rate: Decimal
    gross_amount: Decimal
    transport_cost: Decimal
    lr_no: str | None
    driver_name: str | None
    vehicle_no: str | None
    owner_name: str | None
    hsn_code: str | None
    tcs_amount: Decimal
    add_topay: Decimal
    less_topay: Decimal
    net_amount: Decimal
    po_no: str | None
    branch: str | None
    created_at: datetime
    updated_at: datetime

File: backend-spring/src/main/java/com/nearkart/entity/SaleEntry.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "sale_entries")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class SaleEntry {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "company_id")
    private UUID companyId;

    @Column(name = "buyer_id", nullable = false)
    private UUID buyerId;

    @Column(name = "product_id", nullable = false)
    private UUID productId;

    @Column(name = "invoice_no", nullable = false)
    private String invoiceNo;

    @Column(name = "sale_date")
    private LocalDate saleDate;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal quantity;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal rate;

    @Column(name = "gross_amount", nullable = false, precision = 14, scale = 2)
    private BigDecimal grossAmount;

    @Column(name = "transport_cost", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal transportCost = BigDecimal.ZERO;

    @Column(name = "lr_no")
    private String lrNo;

    @Column(name = "driver_name")
    private String driverName;

    @Column(name = "vehicle_no")
    private String vehicleNo;

    @Column(name = "owner_name")
    private String ownerName;

    @Column(name = "hsn_code")
    private String hsnCode;

    @Column(name = "tcs_amount", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal tcsAmount = BigDecimal.ZERO;

    @Column(name = "add_topay", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal addTopay = BigDecimal.ZERO;

    @Column(name = "less_topay", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal lessTopay = BigDecimal.ZERO;

    @Column(name = "net_amount", nullable = false, precision = 14, scale = 2)
    private BigDecimal netAmount;

    @Column(name = "po_no")
    private String poNo;

    private String branch;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "buyer_id", insertable = false, updatable = false)
    private Buyer buyer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", insertable = false, updatable = false)
    private Product product;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now; updatedAt = now;
        if (saleDate == null) saleDate = LocalDate.now();
    }

    @PreUpdate
    protected void onUpdate() { updatedAt = OffsetDateTime.now(); }
}
